### ISSUE TITLE
Feature Request: Contraint Modelling

### ISSUE THREAD

**Issue by giannissc** [GraphiteEditor/Graphite#216]
Coming from the world of CAD I would really like it if constraints were available for modelling. They can be really useful when changing a design but wanting some features to be connected a certain way. The alternative way of doing this without constraints would be to scale every component independently and moving them around until thing look right.

I haven't seen in into any other program (other than point coincident contraint) and I think it would be really useful.

**Comment by TrueDoctor** [comment]
Could you maybe give a small example (e.g. screenshot)?

**Comment by giannissc** [comment]
I don't have an example specific to 2D graphics (since none implement constraint but coming from a 3D cad environment one typically uses 2D sketches to implemnt certain aspect of a design 

![image](https://user-images.githubusercontent.com/20277283/122239763-1d40f500-ceb9-11eb-9f99-e01f75438d22.png)

On the design below we can see various contraints in place (point coincident, horizontal, vertical contratins, symmetrical contraints, perpendicular constraints and midpoint constraints)

**Comment by giannissc** [comment]
No if I were to change some aspect of this design all the contraint would be checked and the constraint solver would try and upheld them by changing some aspect of the design

**Comment by giannissc** [comment]
A could program to play around with to get the hang of constraint and see how they might by useful would be solvespace https://github.com/solvespace/solvespace/

**Comment by giannissc** [comment]
Finally another example could be a simpe clock design like so:

![image](https://user-images.githubusercontent.com/20277283/122246193-40ba6e80-cebe-11eb-8db7-5b377bf6cf23.png)

Where I could constraint the minute hand to be shorter than the radius of the clock and the hour hand to be shorter by some ratio than the minute hand and then modify my design until it is what I want

**Comment by giannissc** [comment]
I hope that this is enought! I hope you like the idea and it fits into what you have envisioned for Graphite :)

**Comment by TrueDoctor** [comment]
Yeah, that is certainly very helpful @Keavon probably has more insight into to what extent that is already planned but these examples make your point very clear.

**Comment by TrueDoctor** [comment]
I would imagine this is also somewhat similar to how skeletal animation software does things with the concept of having something like joints

**Comment by giannissc** [comment]
I haven't done any skeletal animations before but I would imagine that it would be useuful in that context as well! Nonetheless, I don't think that is a must have but rather a nice to have feature that once you get used to you won't want to go back 😋

**Comment by Keavon** [comment]
I haven't replied yet to #217 and this one because I want to put in the time to write out my thoughts in detail. But very briefly, I am very excited that you have started a discussion about this in a venue where we can continue discussing it until eventually implementing the results of the proposals. The concept of CAD-like constrain systems, or more broadly the procedural/parametric approach used by CAD software, is something I have been wanting to learn more about to integrate those concepts into Graphite's procedural system. I see many uses for such a system, and probably this and #217 can be implemented in the same system to be useful for UIs and everything else. I'll try to get the time in the next few days to read through this thread and comment about what I imagine is possible and what needs further investigation and thought.

**Comment by giannissc** [comment]
A good review of the space I have found is found in [1-3]. [1] is more relevant as it describes the problem of constraints in 2D. Early history of contraint solving is related to CAD modelling and the main people you should look research from are Glenn A. Kramer [4-6], Ioannis Fudos [7], Christoph M. Hoffmann [2,7]. This is from a quick research I did I havent read through them just skimmed them. I will update as I read more about them :) 

References:
[1](https://github.com/GraphiteEditor/Graphite/files/6676933/41d4.content.pdf)
[2](https://github.com/GraphiteEditor/Graphite/files/6676971/021001_1.pdf)
[3](https://github.com/GraphiteEditor/Graphite/files/6677016/134.pdf)
[4](https://github.com/GraphiteEditor/Graphite/files/6676979/1-s2.0-000437029290012M-main.pdf)
[5](https://github.com/GraphiteEditor/Graphite/files/6676981/112515.112566.pdf)
[6](https://github.com/GraphiteEditor/Graphite/files/6676982/AAAI90-106.pdf)
[7](https://github.com/GraphiteEditor/Graphite/files/6677010/TOG97.pdf)
